File: Cpp/Continuous Subarray Sum II.cpp
#include <vector>

class Solution {
public:
    /**
     * @param A an integer array
     * @return  A list of integers includes the index of
     *          the first number and the index of the last number
     */
    vector<int> continuousSubarraySumII(vector<int>& A) {
        // Write your code here
        vector<int> res;
        //if (A.empty()) return res;
        int len = A.size(), sum = 0;
        for (int i = 0; i < len; i ++) {
            sum += A[i];
        }
        int lmax = 0, rmax = 0, cnt = A[0], maxv = A[0], pos = 0;
        for (int i = 1; i < len; i ++) {
            if (cnt < 0) {
                cnt = 0;
                pos = i;
            }
            cnt += A[i];
            if (cnt > maxv) {
                maxv = cnt;
                lmax = pos;
                rmax = i;
            }
        }
        cnt = A[0], pos = 0;
        int lmin = 0, rmin = 0, minv = A[0];
        for (int i = 1; i < len; i ++) {
            if (cnt > 0) {
                cnt = 0;
                pos = i;
            }
            cnt += A[i];
            if (cnt < minv) {
                minv = cnt;
                lmin = pos;
                rmin = i;
            }
        }
        if (maxv > sum - minv || (rmin - lmin + 1 == len)) { // all nums are negative
            res.push_back(lmax);
            res.push_back(rmax);
        } else {
            if (++rmin == len) rmin = 0;
            if (--lmin < 0) lmin = len - 1;
            res.push_back(rmin);
            res.push_back(lmin);
        }
        return res;
    }
};

File: Java/Continuous-Subarray-Sum-II.java
public class Solution {
    /**
     * @param A an integer array
     * @return  A list of integers includes the index of the first number and the index of the last number
     */
    public ArrayList<Integer> continuousSubarraySumII(int[] A) {
        // Write your code here
        int leftm = 0, rightm = 0, cur = -1, maxv = Integer.MIN_VALUE, sum = 0, pos = 0;
        for (int i = 0; i < A.length; i ++) {
            sum += A[i];
            if (cur < 0) {
                pos = i;
                cur = A[i];
            } else {
                cur += A[i];
            }
            if (cur > maxv) {
                maxv = cur;
                leftm = pos;
                rightm = i;
            }
        }
        int leftmin = 0, rightmin = 0, minv = Integer.MAX_VALUE;
        pos = 0;
        cur = 1;
        for (int i = 0; i < A.length; i ++) {
            if (cur > 0) {
                pos = i;
                cur = A[i];
            } else {
                cur += A[i];
            }
            if (cur < minv) {
                minv = cur;
                leftmin = pos;
                rightmin = i;
            }
        }
        ArrayList<Integer> res = new ArrayList<Integer>();
        if (maxv > sum - minv || (rightmin - leftmin + 1 == A.length)) {
            res.add(leftm);
            res.add(rightm);
        } else {
            rightmin ++;
            leftmin --;
            if (rightmin == A.length) rightmin = 0;
            if (leftmin < 0) leftmin = A.length - 1;
            res.add(rightmin);
            res.add(leftmin);
        }
        return res;
    }
}